Histogram chart for job-specific metrics

adamlogic opened the pull request:
Work-in-progress while I explore some possibilities.

adamlogic commented:
Here's my first exploration here - using a bubble chart to visualize the histogram data for a job.

<img width="1100" alt="CleanShot 2022-08-12 at 15 36 55@2x" src="https://user-images.githubusercontent.com/372/184430935-2afaa8a7-2688-46d1-8be0-a84e6591c192.png">

<img width="1108" alt="CleanShot 2022-08-12 at 15 36 49@2x" src="https://user-images.githubusercontent.com/372/184430941-f32ec9cf-0fc5-4ec4-98cb-506fe2817d6f.png">

Curious for feedback before I move forward.

adamlogic commented:
The other thing I was going to check out was this [box plot plugin](https://github.com/sgratzl/chartjs-chart-boxplot) for Chart.js. LMK if you think I should go down that route.

mperham commented:
I’m not sure about the y axis. I’m not clear how to identify it as logarithmic, not linear.

the bubbles have a lot of visual noise, maybe the box plot will be cleaner? 🤷🏻‍♂️

adamlogic commented:
Here's an attempt at a box plot chart. Y-axis is milliseconds.

<img width="1112" alt="CleanShot 2022-08-13 at 08 39 30@2x" src="https://user-images.githubusercontent.com/372/184494448-5a016dbe-cfcd-41b7-98bf-f97c1d0878e6.png">

<img width="1131" alt="CleanShot 2022-08-13 at 08 39 18@2x" src="https://user-images.githubusercontent.com/372/184494450-34b33ce3-9a22-4c70-9bda-e6a0add0de45.png">

And here's another version of the bubble chart with a linear y-axis (same y-axis as the box plot above).

<img width="1076" alt="CleanShot 2022-08-13 at 08 40 42@2x" src="https://user-images.githubusercontent.com/372/184494483-a487b2e9-be0a-4506-9280-e428e05cf694.png">

<img width="1098" alt="CleanShot 2022-08-13 at 08 40 30@2x" src="https://user-images.githubusercontent.com/372/184494505-9f463faf-3f55-4aef-8b59-bed8b39a7f1e.png">

I personally find the logarithmic axis more readable and a better visual aesthetic, and I think the bubbles convey more information than the box plot. My favorite amongst these is the first version (bubbles with logarithmic axis), but I'll roll with the one you prefer. Or we can keep exploring if you have more ideas.

adamlogic commented:
I deployed this branch to our staging environment for Rails Autoscale to get some screenshots of real data.

![CleanShot 2022-08-13 at 08 54 59@2x](https://user-images.githubusercontent.com/372/184495024-15fe2bba-d1dd-47ba-a46b-469788e53b5f.png)

![CleanShot 2022-08-13 at 08 55 15@2x](https://user-images.githubusercontent.com/372/184495026-70db2f7a-55d6-467f-a60e-4f1e467bb90b.png)

![CleanShot 2022-08-13 at 08 55 32@2x](https://user-images.githubusercontent.com/372/184495027-87aa2956-1d33-49b0-b434-fcb6ec70e309.png)

adamlogic commented:
And just for fun, here's the metrics overview page.

![CleanShot 2022-08-13 at 08 59 00@2x](https://user-images.githubusercontent.com/372/184495263-e22959c5-8551-42f2-8b57-c3717d8e31c3.png)

mperham commented:
The box plot is too visually “heavy” with the box filled in. And if the box is just outlined, that would make the dot in the middle pop much better. I do like the bubbles though.

adamlogic commented:
Here it is with the box outlined. Note that these examples look a lot different because we have some extreme outliers now—these happened by accident when I closed my laptop for a bit while jobs were running.

![CleanShot 2022-08-13 at 16 12 16@2x](https://user-images.githubusercontent.com/372/184509351-28dd62f6-538c-4ec7-bc0f-1c7c45f2dabe.png)

![CleanShot 2022-08-13 at 16 12 37@2x](https://user-images.githubusercontent.com/372/184509352-cebc2da4-f7d8-4208-b30c-a7baa6cd4a42.png)

I will say that the bubble chart with the linear scale does a much better job of showing just how far these outliers are from the rest, which I think is helpful. The bubbles at the bottom are all muddied together, but that's not the interesting data in the chart.

The linear scale also aligns with the scale we're using on the metrics overview chart (especially once I change the metric from milliseconds to seconds and add the y-axis label). I think there's benefit to having consistency between the two pages.

I'm definitely still pro-bubbles, but I'm starting to lean toward the linear scale vs. logarithmic.

mperham commented:
I'm ok with letting you drive the design. If you think bubbles work best, ship it. Others can comment or propose improvements over time.

adamlogic commented:
> I'm ok with letting you drive the design.

Famous last words. 😂 

I'm happy with the current state of this PR, so I'm marking it ready for review. Here's where I landed:

<img width="1298" alt="CleanShot 2022-08-15 at 08 21 16@2x" src="https://user-images.githubusercontent.com/372/184634255-ecd9d1df-362b-4efb-990c-1bbaaf350abd.png">

![CleanShot 2022-08-15 at 08 23 12@2x](https://user-images.githubusercontent.com/372/184634446-89d8a74a-8ca9-41be-94ea-0e9cc35a2b4c.png)

![CleanShot 2022-08-15 at 08 22 30@2x](https://user-images.githubusercontent.com/372/184634451-df27699e-3b49-477e-b342-40d3ffae1f59.png)

mperham commented:
Looks nice so far, I'll test it soon. Be sure to check dark mode.

mperham commented:
Total and Average Execution Time should probably have three digits, i.e. "421 sec" or "4.21 sec", so the math is clear, otherwise you get strange calculations like Processed = 7, TET = 2, Average = 0.42 (???)

adamlogic commented:
I just changed the rounding so all of the seconds measurements on these pages are rounded to two decimals for consistency. Also, here's dark mode:

<img width="1215" alt="CleanShot 2022-08-15 at 15 10 33@2x" src="https://user-images.githubusercontent.com/372/184701351-0c7119ba-36d3-4e1d-b947-7d7107c1b0ca.png">

mperham commented:
Nice. Can you verify the page looks good with no data and I'll merge?

mperham commented:
What do you think about two level headers:

h1 Metrics / OrderJunkJob

h2 Histogram Summary

h2 Performance over Time

Where the h1 is providing breadcrumbs and Metrics can link to the previous page (i.e. metrics home).

mperham commented:
What do we do with the data table? Doesn't seem super useful, especially if it shows rows with all zeros.

adamlogic commented:
> What do you think about two level headers

👍 I agree. I don't think there's any precedent for subheaders in the UI, so I'll do my best to make it work visually.

> What do we do with the data table? Doesn't seem super useful, especially if it shows rows with all zeros.

I vote to keep it. The data in the table is not reflected in the charts above, so it _could_ be helpful to dig into specifics at a given point in time.

Not a strong opinion, though. Happy to drop it if that's your preference.

![CleanShot 2022-08-16 at 12 36 17@2x](https://user-images.githubusercontent.com/372/184932678-c20fedaa-7eea-47b4-b96f-a70760e1f55f.png)

adamlogic commented:
Here are the new headers for the metrics pages. Showing the dashboard as well to see that the header style matches (even though all the other pages use an `h3` instead of `h1`).

https://user-images.githubusercontent.com/372/184937807-0b93d945-dca2-4fc8-86a0-4f2de050f63b.mp4

And here's how it looks with no data at all.

![CleanShot 2022-08-16 at 12 54 36@2x](https://user-images.githubusercontent.com/372/184937867-ae6e6acb-52b3-4923-8bb6-e231c3813587.png)

mperham commented:
I don't think you need the Metrics header on the Metrics home page. It's useful on the Metrics detail page to link back to the Metrics home. It's not a bad idea for all detail pages: Queues > default, Batches > $BID, Retries > $JID, Limits > $NAME, etc. Let's avoid scope creep and make that a TODO in the future for now.

mperham commented:
If the data is completely empty, I think you should display a separate "No data found" message like on the other existing pages. For instance, retries or scheduled data.

adamlogic commented:
Removed the metrics header:

<img width="1180" alt="CleanShot 2022-08-16 at 13 53 05@2x" src="https://user-images.githubusercontent.com/372/184946510-aa4afce5-c998-4eba-9ff5-22c6d468413e.png">

Added "no job metrics" message:

<img width="1131" alt="CleanShot 2022-08-16 at 13 53 22@2x" src="https://user-images.githubusercontent.com/372/184946552-cdd6df74-8b82-4364-87fd-4d13d196f363.png">
